### PR TITLE
Implement FilterAndRerank for policy prediction

### PR DESCRIPTION
**Pull request by varunshankar**
**Proposed changes**:
- Implements https://rasahq.atlassian.net/browse/ENG-87
- Adds a hook to get filtered and reranked policy predictions.

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/core/policies/ensemble.py [hunk 1]
```diff
@@ -275,6 +276,9 @@ def _best_policy_prediction(
             for prediction in predictions:
                 prediction.probabilities[index_of_rejected_action] = 0.0
 
+        predictions = plugin_manager().hook.filter_and_rerank_actions(
```
Comment (d61h6k4) [inline rasa/core/policies/ensemble.py:279]: To use only the first result, you can mark the hook with `@hookspec(firstresult=True)` 
https://pluggy.readthedocs.io/en/stable/#first-result-only

**Review by d61h6k4 — CHANGES_REQUESTED** [review]

File: rasa/core/policies/ensemble.py (hunk 1 above)
Comment (twerkmeister) [reply rasa/core/policies/ensemble.py:279]: Also we should cover the case when people are not using rasa pro thus the hook returning `None` (I believe). In this case the original predictions should not be overwritten

**Review by twerkmeister — CHANGES_REQUESTED** [review]

**Review by twerkmeister — APPROVED** [review]
Looks good and can probably work this way. We really need to test this though with some bot in an integrated fashion.

**Review by d61h6k4 — APPROVED** [review]
👍🏻

**Pull request merged**